core/src/test/java/org/elasticsearch/index/replication/IndexLevelReplicationTests.java

```diff
@@ -156,4 +165,76 @@ public void testCheckpointsAdvance() throws Exception {
         }
     }
 
+    /**
+     * test document failures (failures after seq_no generation) are added as noop operation to the translog
+     * for primary and replica shards
+     */
+    public void testDocumentFailureReplication() throws Exception {
+        IndexMetaData metaData = buildIndexMetaData(1);
+        final ReplicationGroup replicationGroupWithDocumentFailureOnPrimary = new ReplicationGroup(metaData) {
+            @Override
+            protected EngineFactory getEngineFactory(ShardRouting routing) {
```

Comment:
can we encapsulate this in a utility class (similar to what i did)? there's a lot of craft here.